[Kernel] Correctly invoke prefill & decode kernels for cross-attention (towards eventual encoder/decoder model support)

afeldman-nm opened the pull request:
This PR is a step towards encoder/decoder model support. This PR modifies the xFormers backend* such that (1) the attention impl can implement cross-attention, and (2) the attention metadata data structure can represent the necessary metadata for invoking cross-attention. 

\* FlashAttention backend support for encoder/decoder models is left as future work

----------

A quick overview of the plan for supporting encoder/decoder models in vLLM:
* Architectural assumptions:
    * The encoder/decoder model comprises one non-autoregressive encoder module and one autoregressive decoder module.
    * A single inference call to the model consumes an encoder prompt and a decoder prompt. The model output is the result of decoder inference against the decoder prompt, conditional on the encoder hidden states which result from applying the encoder to the encoder prompt. The encoder hidden states are *not* part of the overall model output
    * Thus, encoder inference is a prerequisite for decoder inference. The decoder consumes encoder hidden states via cross-attention, which is not present in decoder-only models.
    * It is assumed that these architectural details are handled inside the model definition; however, to support the inference process for such models, vLLM core must be changed to accommodate *cross-attention*
* Encoder/decoder inference process & cross-attention:
    * **Prefill phase:** (1) Non-autoregressive encoder inference yields encoder hidden states in a single pass; no KV caching occurs. (2) decoder prefill yields first-token-prediction & cached KVs. Within the decoder, cross-attention layers cache the KVs derived from encoder hidden states:
    
        * *Key_{cross-attn, layer-n} = W_{K, cross-attn, layer-n} x (Encoder hidden states)*
        * *Value_{cross-attn, layer-n} = W_{V, cross-attn, layer-n} x (Encoder hidden states)*
    
        * Note that all cross-attention layers consume the same encoder hidden states; however each cross-attention layers' keys and values differ because each layer has unique *W_{K, cross-attn, layer-n}* and *W_{V, cross-attn, layer-n}*. **Therefore, the cross-attention KV cache must store KVs for each decoder layer, even though these KVs are all derived from a single set of encoder hidden states.**
    
        * Note that self-attention layer behavior is unchanged compared to what it would be in a decoder-only model (cache KVs computed from the previous decoder layer outputs.)
    
    * **Decode phase:** during each iteration of the autoregressive decode process,
        * Each self-attention layer appends the last predicted token's KVs to the KV cache, and then utilizes cached KVs for next-token prediction (again, this is unchanged compared to a decoder-only model)
        * Each cross-attention layer has *read-only access* to cross-attention KVs, to use for next-token prediction. The cross-attention KV cache is never modified after prefill

-----

To implement the above encoder/decoder inference process, the following functionality will be added to vLLM over the course of multiple PRs:
1. Support cross-attention KV cache & memory management (allocate/swap/free) in block manager
    * See PR: https://github.com/vllm-project/vllm/pull/4837
2. **(This PR) Invoke cross-attention *operation* via the Attention wrapper & Attention metadata data structure**
3. Modify ModelRunner to construct input tensors & Attention metadata structures for cross-attention
4. Small changes to LLM engine & scheduler so that vLLM requests can include an encoder input prompt

Note 1: because this PR makes an incremental contribution (cross-attention KV-caching and memory management), this PR will not enable end-to-end encoder/decoder support (this will rely on later PRs.)

Note 2: the best effort is being made to ensure that encoder/decoder models are compatible with existing vLLM features. At this time, encoder/decoder models are unlikely to be compatible with the following vLLM features:
* Speculative decoding
* Chunked prefill
* Automatic prefix caching
* Sliding window
* Flash attention
* CUDA graph

INCREMENTAL FIX TOWARDS #187 

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

sroy745 commented on vllm/attention/backends/xformers.py:
```diff
@@ -116,65 +145,196 @@ def __post_init__(self):
 
+    @property
+    def has_valid_cross_attn_metadata(self):
+        # No cross-attention metadata is present whatsoever
```
can we return (self.cross_seq_lens is not
                 None) and (self.cross_slot_mapping is not
                            None) and (self.cross_block_tables is not None) here and in the do_cross_attn setter function assert it?

afeldman-nm replied:
Just pushed a fix for this.

sroy745 commented on vllm/attention/backends/xformers.py:
```diff
@@ -116,65 +145,196 @@ def __post_init__(self):
         self.attn_bias: Optional[List[AttentionBias]] = None
 
+    @property
+    def has_valid_cross_attn_metadata(self):
```
In this method are we trying to validate if all of the cross attention metadata values are set? In that case can we rename to is_all_cross_attn_metadata_set?

afeldman-nm replied:
Yes, just pushed fix.

sroy745 commented on tests/kernels/test_self_and_cross_attn.py:
```diff
@@ -0,0 +1,1250 @@
+        self_prefill_packed_actual_output.view_as(
+            self_prefill_packed_ideal_output))
+
+    cross_prefill_packed_actual_output: torch.Tensor = run_cross_attention_test(
```
Can you please clarify what we are trying the verify in this run_cross_attention_test? I was thinking we will have 3 things to verify here a) the encoder attention computation what we are doing and verifying right before this in L1202 (b) verify the decoder attention what we are doing in L1238 and c) verify the cross attention between the decoder attention and the encoder/cross attention what we are doing  L:1247.  So wondering which portion of attention computation this is verifying?

afeldman-nm replied:
Hi @sroy745 , there is a effectively a "grid" of test-case combinations. You are correct that one dimension of this grid is the (a)/(b)/(c) you have specified in your comment above, i.e. "type of attention".

However, another dimension of the test-case grid is prefill-phase vs decode-phase. This is because (1) prefill and decode use different kernels and (2) these kernels rely on different AttentionMetadata attributes being set & with different attribute values.

So the space of tests that needs to be considered is (prefill-phase, decode-phase) x (encoder test, decoder self-attention test, encoder/decoder cross-attention test); the only combination of test configurations which we skip is decode-phase + encoder test, because (under my implementation) the encoder inference is entirely completed during prefill, so there is no encoder inference during decode phase. Therefore there are 5 test conditions.

In order to make clearer how this is accomplished in my code, I recently made the following changes:
* run_self_attention_test() is renamed to run_encoder_or_decoder_self_attention_test(), so it is clear that this helper function may be used *either* to test encoder attention, or decoder self-attention (the attention metadata configuration is almost identical for both)
* run_cross_attention_test() is renamed to run_encoder_decoder_cross_attention_test(), so it is clear that this helper function may be used *only* to test encoder/decoder cross-attention
* I had actually forgotten to add a unit test for the prefill-phase & **encoder** test scenario; I just added it: test_encoder_attention()

If you look at test_enc_dec_self_and_cross_attention_prefill_decode_phases(), this test considers the following combination of scenarios:

(prefill-phase, decode-phase) x (decoder self-attention, encoder/decoder cross-attention)

The prefill-phase tests populate the self-attention and cross-attention KV caches, which are then re-used by the subsequent decode-phase tests.

The new encoder attention test I added, test_encoder_attention(), tests only the prefill-phase/encoder scenario. This scenario does not utilize the KV cache at all.

A note on terminology -

I have in the past used the term "cross-attention" interchangeably with "encoder/decoder attention", and I have used "self-attention" as if it were an adjective that describes both encoder attention and decoder self-attention (since both types of attention derive Q/K/V solely from outputs of the previous attention layer.) However based on your feedback, I think my terminology was possible unclear/incorrect. So I have adjusted to use the following terminology:

* The variety of attention that happens *inside* the encoder == "Encoder attention"
* The variety of attention where the decoder layer Q/K/V are all derived from the outputs of the previous decoder layer == "Decoder self-attention"
* The variety of attention where the decoder layer Q is derived from the previous layer output, while K & V are either derived from encoder hidden states or reused from cache == "Encoder/decoder cross attention"

afeldman-nm committed:
Merge branch 'infra_enc_dec_block_manager' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_block_manager

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_block_manager_reviews

afeldman-nm committed:
return output of SequenceGroup constructor

afeldman-nm committed:
capitalize constants

afeldman-nm committed:
refactored swap-block-table functionality

afeldman-nm committed:
Refactored block manager + enc dec + unsupported feature checks into utils

afeldman-nm committed:
removed circular import

afeldman-nm committed:
apparently isort has to run last?

afeldman-nm committed:
slight name change

afeldman-nm committed:
merge

afeldman-nm committed:
wip merge

afeldman-nm committed:
fixed utils to correctly handle encoder/decoder unsupported scenarios

afeldman-nm committed:
formatting

afeldman-nm committed:
Merge branch 'infra_enc_dec_block_manager' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
renamed xformers metadata is_cross_attn to is_encoder_decoder_attn

afeldman-nm committed:
wip getting tests to pass after merge

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
passing tests; formatting

afeldman-nm committed:
removed overprovisioning from make_block_tables_slot_mapping()

afeldman-nm committed:
comments'

afeldman-nm committed:
clarified block table address formula

afeldman-nm commented on tests/layer/test_self_and_cross_attn.py:
```diff
@@ -0,0 +1,1256 @@
+
+    The first block is at
+
+    block_base_addr + sum(num_blocks_list) * 2 - 1
```
Ahh. When I first wrote this code I grossly overprovisioned the blocktable size in order to decouple silly off-by-one errors in the block table size computation from actual functional issues with the attention kernel invocation.

However, now that the attention kernel invocation passes tests, there is no reason for this overprovisioning.

So I went back and removed the overprovisioning from the blocktable size computation:

```
    # Provision minimum number of KV cache blocks
    num_blocks_list = [
        num_tokens_to_min_blocks(num_tokens, block_size)
        for num_tokens in seq_lens
    ]
    
    ...

    # Compute uppermost address of block table
    total_cache_blocks = sum(num_blocks_list)
    block_base_idx = block_base_addr + total_cache_blocks
    max_block_idx = block_base_idx
```

And then I added the following explanatory comment to the beginning of make_block_tables_slot_mapping():

```
    Construct fake block tables & slot mappings.

    For a sequence with num_tokens tokens the minimum number
    of required KV cache blocks is

    num_blocks = (num_tokens + block_size) // block_size

    Then the minimum KV cache size in blocks is

    total_cache_blocks = sum(num_blocks for all seqs) 

    Then, the blocktable mapping counts downward from

    block_base_addr + total_cache_blocks

    to

    block_base_addr
```

afeldman-nm commented on tests/layer/test_self_and_cross_attn.py:
```diff
@@ -0,0 +1,1256 @@
+
+    scale = float(1.0 / (head_size**0.5))
+    attn_backend = make_backend(backend_name)
+    attn = make_attention(num_heads, head_size, scale)
```
Yes, fixed in most recent push.

afeldman-nm committed:
wip changing cross attention flag

afeldman-nm committed:
yapf fix

afeldman-nm committed:
yapf fix

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_block_manager_reviews

afeldman-nm committed:
upstream merge

afeldman-nm committed:
fix formatting issue

afeldman-nm committed:
formatting

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_block_manager_reviews

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_block_manager_reviews

afeldman-nm committed:
Merge branch 'infra_enc_dec_block_manager' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
passing tests with new attention type enum

afeldman-nm committed:
formatting

afeldman-nm committed:
wip encoder test

afeldman-nm committed:
first pass at encoder attention test

afeldman-nm committed:
wip encoder attention test

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
encoder attention test passes!

afeldman-nm committed:
formatting

afeldman-nm committed:
encoder test arguments

afeldman-nm committed:
type hints; formatting

afeldman-nm committed:
merge in review work

afeldman-nm committed:
typo

afeldman-nm committed:
changed helper function naming convention

afeldman-nm committed:
check we are not testing decode-phase/encoder attention

afeldman-nm committed:
refactoring

afeldman-nm committed:
formatting

afeldman-nm committed:
removing unnecessary check

afeldman-nm commented on vllm/attention/backends/xformers.py:
```diff
@@ -253,33 +413,61 @@ def forward(
-        key = key[:num_prefill_tokens]
-        value = value[:num_prefill_tokens]
+
+        if not do_cross_attn and key is not None and value is not None:
```
You are right, it is unnecessary. Pushed a fix

afeldman-nm commented on vllm/attention/backends/xformers.py:
```diff
@@ -374,8 +575,12 @@ def _run_memory_efficient_xformers_forward(
+                if attn_metadata.is_cross_attn:
+                    attn_bias = BlockDiagonalMask.from_seqlens(
+                        attn_metadata.seq_lens, attn_metadata.cross_seq_lens)
+                else:
```
Correct, I fixed the masking logic to be as follows (recall from one of my other comments that I replaced attn_metadata.is_cross_attn with attn_metadata.attention_type which is an enum-type member):

```
        if attn_metadata.attn_bias is None:
            if self.alibi_slopes is None:
                if attn_metadata.attention_type == \
                    AttentionType.ENCODER_DECODER:
                    # Default enc/dec cross-attention mask is non-causal
                    attn_bias = BlockDiagonalMask.from_seqlens(
                        attn_metadata.seq_lens, attn_metadata.cross_seq_lens)
                else:
                    if attn_metadata.attention_type == AttentionType.ENCODER:
                        # Default encoder self-attention mask is non-causal
                        attn_bias = BlockDiagonalMask.from_seqlens(
                            attn_metadata.seq_lens)
                    else:
                        # Default decoder self-attention mask is causal
                        attn_bias = BlockDiagonalCausalMask.from_seqlens(
                            attn_metadata.seq_lens)
                if self.sliding_window is not None:
                    attn_bias = attn_bias.make_local_attention(
                        self.sliding_window)
                attn_metadata.attn_bias = [attn_bias]
            else:
                attn_metadata.attn_bias = _make_alibi_bias(
                    self.alibi_slopes, self.num_kv_heads, query.dtype,
                    attn_metadata.seq_lens)
```

So encoder attention and encoder/decoder cross-attention will now default to non-causal, while decoder attention defaults to causal. Both may be overridden with a custom attention bias.

afeldman-nm committed:
unit test for encoder/decoder+chunked prefill non-support; added attention utils file for error strings

afeldman-nm committed:
explanatory comment

afeldman-nm committed:
refactoring

afeldman-nm commented on vllm/attention/backends/xformers.py:
```diff
@@ -253,33 +413,61 @@ def forward(
-        assert key.shape[0] == num_prefill_tokens + num_decode_tokens
-        assert value.shape[0] == num_prefill_tokens + num_decode_tokens
+
+        assert do_cross_attn or (key.shape[0]
```
Currently the plan is that encoder/decoder does not support chunked prefill and therefore does not support a mix of prefill and decode tokens.

Just pushed a change which
* Raises a NotImplementedError if the following are both true: (1) the attention backend is performing encoder/decoder cross-attention, which is a clear sign of an encoder/decoder model, and (2) there is a mix of encoder and decoder tokens
* Added a test condition to the test_enc_dec_self_and_cross_attention_prefill_decode_phases() test, which checks that invoking the attention backend for encoder/decoder cross attention & a mix of prefill + decode tokens triggers a NotImplementedError. This test condition could in principle have been a standalone test; however the test setup would be so involved that it is easier to just add this "not implemented" test as an additional assert in test_enc_dec_self_and_cross_attention_prefill_decode_phases()

afeldman-nm committed:
spelling fix

afeldman-nm committed:
rename

afeldman-nm committed:
skip enc/dec tests if HIP

afeldman-nm committed:
formatting

afeldman-nm committed:
Refactored checks into utils file

afeldman-nm committed:
format

afeldman-nm committed:
wip trying to combine attention metadata caches

afeldman-nm committed:
wip trying to merge self/cross caches; trying to fix attn_bias issues; just tried having xformers backend clear mask when changing target attention type

afeldman-nm committed:
wip

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
wip merging attention metadata

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn_valid_md

afeldman-nm committed:
simplied is_all_cross_attn_metadata_set()

afeldman-nm commented on vllm/attention/backends/xformers.py:
```diff
@@ -105,8 +106,36 @@ class XFormersMetadata(AttentionMetadata, PagedAttentionMetadata):
+    _self_cached_prefill_metadata: Optional["XFormersMetadata"] = None
+    _self_cached_decode_metadata: Optional["XFormersMetadata"] = None
+    # Cross-attention prefill/decode metadata cache
+    _cross_cached_prefill_metadata: Optional["XFormersMetadata"] = None
```
WIP

afeldman-nm replied:
Fixed in latest push - now there is just _cached_prefill_metadata and _cached_decode_metadata

afeldman-nm commented on vllm/attention/backends/xformers.py:
```diff
@@ -116,65 +145,196 @@ def __post_init__(self):
+        else:
+            # Cross-attention prefill
+
+            if self._cross_cached_prefill_metadata is not None:
```
WIP

afeldman-nm replied:
Fixed in latest push - now there is just _cached_prefill_metadata and _cached_decode_metadata

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn

afeldman-nm committed:
test: envs.VLLM_ATTENTION_BACKEND

afeldman-nm committed:
formatting'

afeldman-nm committed:
attempted to fix issue whereby selector test doesn't cleanup environment variables

afeldman-nm committed:
(1) In top-level tests utils.py added env var context manager, (2) in tests/kernels added utils.py w/ vLLM backend context manager, (3) all unit tests for backend selector & enc/dec use backend context manager

afeldman-nm committed:
wip tests

afeldman-nm committed:
FIX: test_attention_selector.py was leaking VLLM_ATTENTION_BACKEND values; fixed with backend context manager

sroy745 commented on tests/kernels/test_self_and_cross_attn.py:
```diff
@@ -0,0 +1,1711 @@
+    return (num_tokens + block_size) // block_size
+
+
+def make_block_tables_slot_mapping(block_size: int,
```
can we have this method return the block_table and slot_mappings for the complete sequences and in the caller use this return value to construct the needed block table tensors and mappings as needed. The reason I am suggesting this is that in function enc_dec_cross_attn_setup_reuses_query when we are invoking this method we are assigning slot_mapping to decode_slot_mapping (instead of using the decode_slot_mapping) which is confusing.

afeldman-nm replied:
Thanks! Makes sense. I refactored slot-mapping and block-table construction into multiple steps. 

make_empty_block_tables_tensor() builds an empty block-tables tensor (this models the behavior of ModelRunner in prefill phase)

make_empty_slot_mapping_tensor() builds an empty slot-mapping tensor (this is appropriate for enc/dec cross-attention *decode-phase*, since KV sequence length is static so newly decoded tokens do not require growing the KV cache)

make_block_tables_slot_mapping() returns
* block_tables_tensor # block tables for the full sequences
* slot_mapping_list # slot mapping for the full sequences
* max_block_idx # max address in the block table

split_slot_mapping() splits the full slot_mapping_list into prefill and decode slot mappings 

I added the following explanatory comments:

```
# Build prefill- & decode-phase data structures
# for decoder self-attention. Block tables and
# slot mapping must be in a format compatible
# with KV caching & attention kernels
#
# Prefill-phase:
#
# * Empty block-tables tensor
# * Slot-mapping with entries for prompt tokens
#
# Decode-phase:
# * Block-tables tensor with minimum number of blocks
#   required by total num. tokens in the entirety of all sequences
#   (including both prefill & decode)
# * Slot-mapping with entries for tokens that will be decoded in the
#   current decode iteration

...

# Build prefill- & decode-phase data structures
# for encoder/decoder cross-attention. Block tables and
# slot mapping must be in a format compatible
# with KV caching & attention kernels
#
# Whereas decoder self-attention extracts relationships between
# equal-length Q/K/V sequences, which mutually grow in length
# with each decoded token, cross-attention relates the Q sequence
# - which grows with each new decoded token - to fixed-length
# K and V sequences derived from the encoder hidden states.
#
# Prefill-phase:
#
# * Empty block-tables tensor
# * Slot-mapping with as many entries as there are tokens in the encoder
#   prompt.
#
# Decode-phase:
# * Block-tables tensor with minimum number of blocks to
#   accommodate K & V tensors which are equal in lnegth
#   to the encoder prompt length
# * Empty slot-mapping tensor (since K & V are fixed in size,
#   new decoded tokens are not KV-cached and require no slot-
#   mapping)
```

sroy745 commented on tests/kernels/test_self_and_cross_attn.py:
```diff
@@ -0,0 +1,1711 @@
+        cross_prefill_packed_ideal_output, \
+        cross_decode_packed_ideal_output, \
+        cross_kv_seq_lens, \
+        cross_decode_block_tables, \
```
wondering if we need all 4  cross_decode_block_tables, cross_decode_slot_mapping, cross_prefill_slot_mapping and  cross_prefill_block_tables here? Won't  cross_prefill_slot_mapping and cross_decode_block_tables be sufficient - the first one for populating the kv cache and the other for the cross attention computation with decoder attention?

afeldman-nm replied:
This is a good question. I think you are alluding to the fact that cross_decode_slot_mapping and cross_decode_block_tables are empty tensors, and so can be omitted. While these tensors are empty, nonetheless they are *tensors* as opposed to being None, thus they still have to be constructed and passed as arguments to attention.

I agree that the test code is complicated by the large numbers of arguments and return-values in each function call.

I have remedied this by creating a set of named tuple data structures that encapsulate the arguments and return values.

afeldman-nm committed:
formatting

afeldman-nm committed:
Merge branch 'upstream-main' into backend_context_manager

afeldman-nm committed:
merge upstream main & well-formatted backend override fix

afeldman-nm committed:
formatting

afeldman-nm committed:
removed comment about supported head_size's, which is not relevant under current encoder/decoder test conditions

afeldman-nm committed:
small refactors

afeldman-nm commented on tests/kernels/test_self_and_cross_attn.py:
```diff
@@ -0,0 +1,1691 @@
+
+# If not is_hip(): supported head sizes are [64, 80, 96, 112, 128, 256]
+#
+# TODO: FlashAttention forward only supports head dimension at most 128
```
This is a carry-over from another test where FlashAttention support mattered. As this code is currently only supporting xFormers this comments are unnecessary & I removed this in the latest push.

afeldman-nm commented on tests/kernels/test_self_and_cross_attn.py:
```diff
@@ -0,0 +1,1691 @@
+    else:
+        packed_query, q_start_loc_list = pack_tensor(query, q_seq_lens)
+    packed_key, kv_start_loc_list = pack_tensor(key, kv_seq_lens)
+    packed_value, _ = pack_tensor(value, kv_seq_lens)
```
Removed in latest push

afeldman-nm commented on tests/kernels/test_self_and_cross_attn.py:
```diff
@@ -0,0 +1,1691 @@
+
+
+def run_encoder_or_decoder_self_attention_test(
+        attn: Attention, packed_query: torch.Tensor, packed_key: torch.Tensor,
```
attn is used here:

https://github.com/neuralmagic/nm-vllm/blob/faf9118554e2677634b3b3bec9d25c5506d1a7d5/tests/kernels/test_self_and_cross_attn.py#L1252

afeldman-nm committed:
refactoring

afeldman-nm committed:
make_qkv() tensors are 4D

afeldman-nm commented on tests/kernels/test_self_and_cross_attn.py:
```diff
@@ -0,0 +1,1691 @@
+    actual_max_q_seq_len = max(q_seq_lens)
+    actual_max_kv_seq_len = max(kv_seq_lens)
+
+    query = torch.rand(
```
Fixed in latest push

afeldman-nm committed:
combined seq_start_loc init with cumsum

afeldman-nm committed:
xformers metadata allows unspecified values for most Optional members; xformers forward only enforces non-none query start locs for prefix caching scenario; simplify self/cross attention test metadata building

afeldman-nm commented on tests/kernels/test_self_and_cross_attn.py:
```diff
@@ -0,0 +1,1691 @@
+                                dtype=torch.int32,
+                                device=device)
+
+    torch.cumsum(seq_lens_tensor,
```
Fixed in latest push

afeldman-nm commented on tests/kernels/test_self_and_cross_attn.py:
```diff
@@ -0,0 +1,1691 @@
+        # Prefill: query_start_loc matches seq_start_loc
+        query_start_loc = copy.deepcopy(seq_start_loc)
+        max_query_len = max_seq_len
+    else:
```
There's actually multiple questions here.
* After re-examining both ModelRunner (which is what builds the metadata structures in practice) as well as observing which metadata fields are actually passed to paged attention by the Attention wrapper, it does appear that seq_lens is in fact always equal to the total length of sequence at the present moment, including generated tokens. So we cannot rely on seq_lens being [1,1,1,...] in the decode case. However...
* It turns out that - aside from in the prefix-caching scenario - xFormers makes no use of query_start_loc or max_query_len whatsoever. Thus, I just pushed a change which modifies the unit test helper functions to simply omit these fields when invoking the metadata constructor. So, the lines in question simply go away.

afeldman-nm committed:
refactored slot mapping logic

afeldman-nm commented on tests/kernels/test_self_and_cross_attn.py:
```diff
@@ -0,0 +1,1691 @@
+        num_blocks = num_blocks_list[sdx]
+        block_table = list(
+            range(block_base_idx, block_base_idx - num_blocks, -1))
+        for idx in range(num_tokens - 1):
```
Yes you are right this is much cleaner! Fixed in latest push.

afeldman-nm committed:
formatting

afeldman-nm committed:
selective renaming of cross -> encoder

afeldman-nm committed:
added encoder, enc/dec cross-attention bias members

afeldman-nm committed:
xformers metadata now uses a different attn_bias for self, encoder and cross

afeldman-nm committed:
refactoring

afeldman-nm committed:
wip typing issues

afeldman-nm committed:
added paged attention args collection, conditional on metadata attention type

afeldman-nm committed:
logic to support encoder-specific sequence length usage in xformers

afeldman-nm committed:
formatting

afeldman-nm committed:
test name change; encoder functionality can tolerate being provided with only encoder metadata

afeldman-nm committed:
formatting

afeldman-nm committed:
prefill supports shared metadata structure

afeldman-nm committed:
formatting

afeldman-nm committed:
full generalization of prefill & decode metadata structures

afeldman-nm committed:
renamed metdata caching structure

afeldman-nm committed:
reverted my custom env var patch impl

afeldman-nm committed:
monkeypatch works

afeldman-nm committed:
formatting

afeldman-nm committed:
Merge branch 'backend_context_manager' into infra_enc_dec_cross_attn_monkeypatch

afeldman-nm committed:
wip monkeypatch

afeldman-nm committed:
refactored constants into tests/kernels/utils.py

afeldman-nm committed:
Merge branch 'backend_context_manager' into infra_enc_dec_cross_attn_monkeypatch

afeldman-nm committed:
wip enc/dec monkeypatch integration

afeldman-nm committed:
a refactoring backend override functionality into tests/kernels/utils.py

afeldman-nm committed:
Merge branch 'backend_context_manager' into infra_enc_dec_cross_attn_monkeypatch

afeldman-nm committed:
test rename

afeldman-nm committed:
Comments & type hints

afeldman-nm committed:
Merge branch 'backend_context_manager' into infra_enc_dec_cross_attn

afeldman-nm committed:
Merge branch 'upstream-main' into backend_context_manager

afeldman-nm committed:
small refactors per @sroy745 suggestions

afeldman-nm committed:
Merge branch 'upstream-main' into backend_context_manager

afeldman-nm committed:
Merge branch 'backend_context_manager' into infra_enc_dec_cross_attn

afeldman-nm committed:
merged backend env config

afeldman-nm committed:
merge

afeldman-nm committed:
fixed _get_seq_len_block_table_args() to change behavior based on is_prompt

afeldman-nm committed:
removed inference of encoder metadata attributes; removed guessing of encoder seq len tensor device

afeldman-nm committed:
wip refactoring

afeldman-nm committed:
refactored helper functions into diffferent utils files

afeldman-nm committed:
_ for private functions in test_encoder_decoder_attn

afeldman-nm committed:
_ refactor

afeldman-nm committed:
formatting

afeldman-nm committed:
constructing attn md with minimum number of arguments

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn_reviews

afeldman-nm commented on tests/kernels/test_self_and_cross_attn.py:
```diff
@@ -0,0 +1,1711 @@
+            self_prefill_slot_mapping,
+            is_encoder_only_test=False,
+            cross_seq_lens=cross_kv_seq_lens,
+            cross_block_tables=cross_prefill_block_tables,
```
Per I discussion on Slack: I wrote the test in such a way as to mirror what goes on inside in the model forward() method, specifically the way in which the forward() method must utilize the single AttentionMetadata structure passed in by ModelRunner for 2 or 3 different types of attention.

afeldman-nm committed:
merge

afeldman-nm committed:
Formatting

afeldman-nm committed:
typing and formatting

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
refactored block table/slot mapping construction process for decoder into two steps

afeldman-nm committed:
finished breaking block table/slot mapping construction into steps; formatting

afeldman-nm committed:
slight refactor

afeldman-nm committed:
refactored encoder test into the cross-attention test

afeldman-nm committed:
slight refactoring

afeldman-nm committed:
QKVInputs and PackedQKVInputs named tuple integration to simplify test logic

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
refactoring

afeldman-nm committed:
format

afeldman-nm committed:
yapf fix

afeldman-nm committed:
import reorg

afeldman-nm committed:
switched to star import to avoid unsatisfiable formatting constraints

afeldman-nm committed:
progress on memory map structure integration

afeldman-nm committed:
completed integration of KVMemoryMap into tests

afeldman-nm committed:
first step toward QKVO integration into tests

afeldman-nm committed:
wip test params structure integration

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
prephase md struct using test params

afeldman-nm committed:
correctness check helper function

afeldman-nm committed:
wip

afeldman-nm committed:
debugging test params integration

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
passing tests with test params integration

afeldman-nm committed:
format

afeldman-nm committed:
test points and test resources structures integrated

afeldman-nm committed:
formatting

afeldman-nm committed:
first attempt at chunked prefill failure test

afeldman-nm committed:
narrowed the space of test-cases for unsupported scenarios

afeldman-nm committed:
format

afeldman-nm committed:
skeleton of encdec prefix cache failure test; fixed bug where max enc seq len was unused

afeldman-nm committed:
wip prefill test

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
passing prefix cache failure test

afeldman-nm committed:
format

afeldman-nm committed:
type annotations; formatting

afeldman-nm committed:
completely replaced collections.namedtuple with typing.NamedTuple w/ type annotations; formatting

afeldman-nm committed:
removed HIP check; clarified assumptions about supported backends in enc/dec supported feature checks

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
wip

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
wip comments

afeldman-nm committed:
small fix

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
formatting

afeldman-nm committed:
enc/dec test comment updates; some function arg changes; formatting

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
formatting

afeldman-nm committed:
fixed attention selector test to use FLASH_ATTN string constant var in all relevant locations

afeldman-nm committed:
Merge branch 'infra_enc_dec_cross_attn' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
additional commenting & added string constants for other backends

afeldman-nm committed:
Merge branch 'upstream-main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
encoder-only unit test passes

afeldman-nm committed:
refactoring

sroy745 reviewed (commented):
Would be good if you could share some pseudo code about how an encoder-decoder model would orchestrate encoder/decoder and cross-attention computation. I think clarifying that would be good to make sure that the current attention backend changes are able to support the overall flow.

afeldman-nm committed:
merge upstream main; attention type comment

afeldman-nm committed:
comment fixes

afeldman-nm committed:
removed unnecessary tests

afeldman-nm committed:
assert value None-ness matches key None-ness

afeldman-nm committed:
comment fix

afeldman-nm committed:
Remove util fxns & error strings for unneeded tests

afeldman-nm commented on vllm/attention/backends/xformers.py:
```diff
@@ -79,73 +82,165 @@ class XFormersMetadata(AttentionMetadata, PagedAttentionMetadata):
 
+    # Begin encoder attn & enc/dec cross-attn fields...
+
+    # If True, prefill_metadata() and decode_metadata() will return
```
Thanks, resolved in most recent push

afeldman-nm commented on vllm/attention/backends/xformers.py:
```diff
@@ -79,73 +82,165 @@ class XFormersMetadata(AttentionMetadata, PagedAttentionMetadata):
+    # otherwise, self-attention data structures will be returned.
+    _attn_type: AttentionType = AttentionType.DECODER
+
+    # (batch_size,). The "cross-sequence-length" per sequence,i.e. the key/value
```
Thanks, resolved in most recent push

afeldman-nm commented on vllm/attention/backends/xformers.py:
```diff
@@ -256,33 +472,78 @@ def forward(
-        if kv_cache is not None:
+        if key is not None:
+            key = key.view(-1, self.num_kv_heads, self.head_size)
+        if value is not None:
```
Thanks, replaced with asserts in most recent push.

afeldman-nm commented on vllm/attention/backends/xformers.py:
```diff
@@ -256,33 +472,78 @@ def forward(
+
+        # Raise NotImplementedError for unsupported encoder/decoder
+        # scenarios (has no effect on decoder-only models)
+        assert_no_encdec_chunked_prefill_assuming_supported_backend(
```
Agreed, removed in most recent push. Current plan is to catch unsupported encoder/decoder scenarios a few levels up in the encoder/decoder ModelRunner subclass that I am writing.

afeldman-nm commented on vllm/attention/backends/xformers.py:
```diff
@@ -298,6 +559,13 @@ def forward(
                 assert out.shape == output[:num_prefill_tokens].shape
                 output[:num_prefill_tokens] = out
             else:
+                if is_encoder_decoder_metadata_assuming_supported_backend(
```
Agreed, removed in most recent push. Current plan is to catch unsupported encoder/decoder scenarios a few levels up in the encoder/decoder ModelRunner subclass that I am writing.

afeldman-nm commented on tests/kernels/utils.py:
```diff
@@ -1,12 +1,144 @@
+class QKVInputs(NamedTuple):
+    '''
+    Data structure for representing unpacked attention inputs, 
+    query/key/value.
```
Thanks, fixed in most recent push

afeldman-nm commented on tests/kernels/utils.py:
```diff
@@ -1,12 +1,144 @@
+
+        * {query,key,value}: packed (number_of_tokens x num_heads 
+                             x head_size) attention inputs
+        * q_seq_lens: list of query start locations within packed tensor
```
Thanks, fixed in most recent push

afeldman-nm commented on tests/kernels/utils.py:
```diff
@@ -1,12 +1,144 @@
+        * {query,key,value}: packed (number_of_tokens x num_heads 
+                             x head_size) attention inputs
+        * q_seq_lens: list of query start locations within packed tensor
+        * kv_seq_lens: shared list of key/value start locations within
```
Thanks, fixed in most recent push.

afeldman-nm commented on tests/kernels/test_encoder_decoder_attn.py:
```diff
@@ -0,0 +1,1117 @@
+                          -> PhaseTestParameters:
+    (num_heads, head_size, _, batch_size, _, _, max_q_seq_len, _) = test_pt
+
+    scale = test_rsrcs.scale
```
Thanks, fixed in most recent push

afeldman-nm commented on tests/kernels/test_encoder_decoder_attn.py:
```diff
@@ -0,0 +1,1117 @@
+@pytest.mark.parametrize("block_size", BLOCK_SIZES)
+@pytest.mark.parametrize("max_dec_seq_len", MAX_DEC_SEQ_LENS)
+@pytest.mark.parametrize("max_enc_seq_len", MAX_ENC_SEQ_LENS)
+def test_backend_fails_for_chunked_prefill_enc_dec(num_heads: int,
```
Agreed, removed in most recent push. Current plan is to catch unsupported encoder/decoder scenarios a few levels up in the encoder/decoder ModelRunner subclass that I am writing.

afeldman-nm commented on tests/kernels/utils.py:
```diff
@@ -20,3 +152,706 @@ def override_backend_env_variable(mpatch: pytest.MonkeyPatch,
+    Construct fake attention metadata for a given test phase
+    (prefill-phase or decode-phase).
+
+    encoder_test_params and cross_test_params arguments all encoder
```
Woops, "all" should be "allows". Fixed in most recent push.

afeldman-nm commented on vllm/attention/backends/xformers.py:
```diff
@@ -79,73 +82,165 @@ class XFormersMetadata(AttentionMetadata, PagedAttentionMetadata):
+
+    # (batch_size,). The sequence length per sequence. Sequence length means
+    # the computed tokens + new tokens None if it is a decoding.
+    seq_lens: Optional[List[int]] = None
```
The weird ordering of the class members was a hack I made to get around this bug, whereby certain member orderings were not accepted by @dataclass:

https://github.com/python/cpython/issues/102559

Not sure if there is any way around this but I will take a look

afeldman-nm replied:
Hi @sroy745 I think it will be productive if I continue to investigate this bug as a separate effort, for now can we keep the code as-is?

afeldman-nm committed:
merge

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_encoder_only

afeldman-nm committed:
formatting

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
format

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Replace attn_metadata.attention_type and attn_metadata._attn_type with attn_type argument to forward()

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Moved attention type for attn_metadata to attention forward(); added NotImplement failures to backends in non-decoder-only scenarios

afeldman-nm commented on vllm/attention/backends/xformers.py:
```diff
@@ -79,73 +82,165 @@ class XFormersMetadata(AttentionMetadata, PagedAttentionMetadata):
             return None
 
         if self._cached_prefill_metadata is not None:
+            self._cached_prefill_metadata.attention_type = \
```
So, with this question you made me realize a significant opportunity to reorganize the code.

Originally, I was planning that prefill_metadata() and decode_metadata() would return a metadata structure tailored to the attention_type (as we discussed previously.) Thus it made sense to have attention_type be a field in attn_metadata.

However, as discussed, I am now having prefill_metadata() and decode_metadata(), respectively, return a single metadata structure that can be shared between all attention types.

So there is no longer much value in having attention_type be a field of attn_metadata - therefore, I just pushed a change which eliminates the attention_type field. Now, when you call Attention.forward(), you simply pass attn_type as an argument. This controls whether encoder attention, decoder self-attention, or encoder/decoder cross-attention is performed.

The call to Attention.forward now looks like:

```
attn.forward(packed_qkv.query,
                        packed_qkv.key,
                        packed_qkv.value,
                        None,
                        attn_metadata,
                        attn_type=attn_type) # <== Attention type
```

afeldman-nm replied:
I have also tweaked every backend other than XFormers to fail with NotImplemented if any attention type other than decoder-self-attention is specified.

afeldman-nm commented on vllm/attention/backends/xformers.py:
```diff
@@ -154,28 +249,134 @@ def decode_metadata(self) -> Optional["XFormersMetadata"]:
             return None
 
         if self._cached_decode_metadata is not None:
+            self._cached_decode_metadata.attention_type = \
```
As mentioned in my previous answer, I eliminated the attention_type property which addresses this issue.

afeldman-nm committed:
num encoder tokens

afeldman-nm commented on vllm/attention/backends/xformers.py:
```diff
@@ -256,33 +472,78 @@ def forward(
+            # Encoder attention - chunked prefill is not applicable;
+            # derive token-count from query shape & and treat them
+            # as 100% prefill tokens
+            num_prefill_tokens = query.shape[0]
```
Subtly, the number of tokens may be different between the encoder and decoder.

So you are picking up on the lack of a num_encoder_tokens field in the attention metadata structure. As a hack I was using query.shape[0] to get the number of encoder tokens.

I just pushed a fix which adds a num_encoder_tokens field & uses it to get the number of tokens in the encoder case.

afeldman-nm commented on vllm/attention/backends/xformers.py:
```diff
@@ -238,14 +439,29 @@ def __init__(
     ) -> torch.Tensor:
         """Forward pass with xFormers and PagedAttention.
 
+        For decoder-only models: query, key and value must be non-None.
```
Yes, I have put together an example here.

https://gist.github.com/afeldman-nm/d061efe58d4b4f5caa8bc3f9d95a8107

afeldman-nm replied:
https://gist.github.com/afeldman-nm/d061efe58d4b4f5caa8bc3f9d95a8107

afeldman-nm commented:
Hi @sroy745 here you can find pseudocode (based on the vLLM Mixtral implementation) showing the proposal for how to construct an encoder/decoder model

https://gist.github.com/afeldman-nm/d061efe58d4b4f5caa8bc3f9d95a8107

afeldman-nm committed:
merge; resolve conflicts

afeldman-nm committed:
merge conflict; typing; formatting

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

sroy745 commented on tests/kernels/test_encoder_decoder_attn.py:
```diff
@@ -0,0 +1,941 @@
+    of ModelRunner, which constructs a single attention metadata structure for
+    each prefill or decode run. A realistic scenario would rely on the
+    attention backend to utilize the appropriate attention metadata fields
+    according to the value of attn_metadata.attention_type. Thus, this test is
```
nit - update the comment to remove reference to attn_metadata.attention_type?

afeldman-nm replied:
Done, replaced each occurrence with "attn.forward() is passed attn_type=AttentionType.{ENCODER,DECODER,ENCODER_DECODER} in order to configure..."

sroy745 reviewed (commented):
Thanks for addressing the comments and for the example model implementation. Some minor comments. Otherwise LGTM.

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
removed extra line

afeldman-nm committed:
changed nested if/else to elif/else in xformers mask computation code

afeldman-nm committed:
reorganized helper functions that were only being used for testing into tests/kernels/utils.py from vllm/utils.py

afeldman-nm committed:
removed attention_type

afeldman-nm committed:
typing and formatting

afeldman-nm committed:
typing and formatting; fixed escape sequences in comments

afeldman-nm commented on vllm/utils.py:
```diff
@@ -798,6 +799,66 @@ def cuda_device_count_stateless() -> int:
 
     return _cuda_device_count_stateless(envs.CUDA_VISIBLE_DEVICES)
 
+def maybe_make_int_tensor(_list: List[int],
```
Done, fixed in latest push

afeldman-nm commented on vllm/attention/backends/xformers.py:
```diff
@@ -375,18 +668,41 @@ def _run_memory_efficient_xformers_forward(
+                    # Default enc/dec cross-attention mask is non-causal
+                    attn_bias = BlockDiagonalMask.from_seqlens(
+                        attn_metadata.seq_lens, attn_metadata.encoder_seq_lens)
+                else:
```
Done, fixed in latest push

afeldman-nm commented on vllm/attention/backends/xformers.py:
```diff
@@ -236,53 +419,148 @@ def __init__(
-        value = value[:num_prefill_tokens]
+
+        if key is not None and value is not None:
+
```
Done, fixed in latest push

afeldman-nm commented on tests/kernels/utils.py:
```diff
@@ -20,3 +152,730 @@ def override_backend_env_variable(mpatch: pytest.MonkeyPatch,
+    # i.e. 100% prefill tokens or 100% decode tokens
+    #
+    # - If is_prompt, num_prefills_or_decodes is the number of prefills
+    #   and num_prefill_or_decode_tokens is the number of prefill tokens
```
Hello @sroy745 , I think you mean to suggest that `num_prefill_or_decode_tokens` is not longer being defined inside `make_test_metadata()`? But actually, if you look at

https://github.com/neuralmagic/nm-vllm/blob/d23c28466765496049a1696d0a053a0a2505ce9a/tests/kernels/utils.py#L825-L837

You will see

```
    num_prefill_or_decode_tokens = None if seq_lens is None else \
        (sum(seq_lens) if is_prompt else len(seq_lens))
```

i.e. `num_prefill_or_decode_tokens` is being defined and used. So I think it is appropriate to keep these comments, and unless you have strong objections I am going to mark this issue resolved without changing anything about the comments.

afeldman-nm committed:
moved make_tensor_with_pad() helper function back to vllm.utils

afeldman-nm committed:
formatting

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

njhill commented on vllm/attention/backends/xformers.py:
```diff
@@ -154,28 +222,143 @@ def decode_metadata(self) -> Optional["XFormersMetadata"]:
+
+    if attn_type == AttentionType.DECODER:
+        return attn_metadata.attn_bias
+    elif attn_type == AttentionType.ENCODER:
+        return attn_metadata.encoder_attn_bias
+    elif attn_type == AttentionType.ENCODER_DECODER:
+        return attn_metadata.cross_attn_bias
+    else:
+        raise AttributeError(f"Invalid attention type {str(attn_type)}")
```
nit: redundant elif/elses

njhill reviewed (approved):
Thanks @afeldman-nm for all of the great work on this!

Might be good to hold off until tomorrow to merge this, I think @maxdebayser is also reviewing at the moment.

Also wondering if it might be good for @WoosukKwon to have a quick look if he has time, given the attention impl changes.

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
removed redundant elif

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

maxdebayser commented on vllm/utils.py:
```diff
@@ -608,6 +598,16 @@ def make_tensor_with_pad(
     return torch.tensor(padded_x, dtype=dtype, device=device)
 
 
+def str_to_int_tuple(s: str) -> Tuple[int, ...]:
```
nit: it seems that the only thing that changed here is the position of the function in the file. If this could be moved it would make the diff cleaner

afeldman-nm replied:
Thanks, fixed in latest push.

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
reverted unnecessarily vllm/utils.py changes

DarkLight1337 commented:
To speed up the CI queue, I've cancelled the distributed tests for the latest CI run in this PR since they won't pass anyway until #5905 has been merged. Now that it has been merged, please merge `main` into your branch so that the CI can pass once again.

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm commented:
> To speed up the CI queue, I've cancelled the distributed tests for the latest CI run in this PR since they won't pass anyway until #5905 has been merged. Now that it has been merged, please merge `main` into your branch so that the CI can pass once again.

Thanks @DarkLight1337 , just did the merge.

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

WoosukKwon commented:
@afeldman-nm Sorry for the delay. Let me take a look at the PR this afternoon.

WoosukKwon commented on vllm/attention/backends/torch_sdpa.py:
```diff
@@ -158,6 +158,11 @@ def forward(
         """
         assert kv_scale == 1.0
+        if attn_type != AttentionType.DECODER:
+            raise NotImplementedError("Encoder self-attention and " + \
+                                      "encoder/decoder cross-attention " + \
+                                      "are not implemented for " + \
+                                      "TorchSDPABackendImpl")
```
nit: You actually don't need `\` here.

Also, it is recommended to avoid `\`: https://google.github.io/styleguide/pyguide.html#32-line-length

WoosukKwon replied:
ditto for other places.

WoosukKwon commented on vllm/attention/backends/torch_sdpa.py:
```diff
@@ -138,14 +138,14 @@ def __init__(
+            kv_cache: Optional[torch.Tensor],
+            attn_metadata: TorchSDPAMetadata,  # type: ignore
+            kv_scale: float = 1.0,
+            attn_type: AttentionType = AttentionType.DECODER) -> torch.Tensor:
```
nit: This will reduce the number of line changes and make the code look better.
```suggestion
            attn_type: AttentionType = AttentionType.DECODER,
            ) -> torch.Tensor:
```

WoosukKwon commented on vllm/attention/layer.py:
```diff
@@ -83,16 +83,22 @@ def __init__(
+                value: torch.Tensor,
+                kv_cache: Optional[torch.Tensor],
+                attn_metadata: AttentionMetadata,
+                attn_type: AttentionType = AttentionType.DECODER) \
+                    -> torch.Tensor:
```
```suggestion
                attn_type: AttentionType = AttentionType.DECODER,
               ) -> torch.Tensor:
```

WoosukKwon reviewed (commented):
@afeldman-nm Thanks for the PR.

Overall, I'm OK with this change. The code in `xformers.py` is a bit difficult to follow, but it's understandable given the inherent complexity.

The only thing I'd like to ask is to avoid using `\` in the code as guided in https://google.github.io/styleguide/pyguide.html#32-line-length Please use parentheses if the line overflows. Also, please considering adding `,` to the last argument of the methods (particularly in the `forward` method of the attention backends) if that reduces the line changes.

afeldman-nm committed:
Update vllm/attention/backends/torch_sdpa.py

nit: This will reduce the number of line changes and make the code look better.

Co-authored-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

afeldman-nm committed:
Update vllm/attention/layer.py

Co-authored-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'infra_enc_dec_cross_attn_reviews' into infra_enc_dec_cross_attn

afeldman-nm committed:
merged in first pieces of woosuk feedback & latest main; formatting

afeldman-nm committed:
fixed specific point-changes requested by woosuk

afeldman-nm committed:
test_encoder_decoder_attn.py cleanup

afeldman-nm committed:
tests/kernels/utils.py cleanup

afeldman-nm committed:
vllm/attention/backends/abstract.py cleanup

afeldman-nm committed:
vllm/attention/backends/blocksparse_attn.py cleanup

afeldman-nm committed:
vllm/attention/backends/flash_attn.py cleanup

afeldman-nm committed:
cleaning up a number of backends & backends utils.py

afeldman-nm committed:
xformers backend cleanup

afeldman-nm committed:
formatting

afeldman-nm commented:
> @afeldman-nm Thanks for the PR.
> 
> Overall, I'm OK with this change. The code in `xformers.py` is a bit difficult to follow, but it's understandable given the inherent complexity.
> 
> The only thing I'd like to ask is to avoid using `\` in the code as guided in https://google.github.io/styleguide/pyguide.html#32-line-length Please use parentheses if the line overflows. Also, please considering adding `,` to the last argument of the methods (particularly in the `forward` method of the attention backends) if that reduces the line changes.

@WoosukKwon Thanks, I have addressed your feedback (both the specific points of feedback as well as the more general styleguide suggestsions.) 

I agree it is hard to make xformers comprehensible given the complexity, perhaps this can wait for a future PR.

Please let me know if you have any other changes to recommend, or if the fixes look satisfactory.

afeldman-nm commented:
Thanks @sroy745 @maxdebayser @njhill @WoosukKwon @robertgshaw2-neuralmagic for all of your excellent help in landing this PR!
